src/backend/distributed/shardsplit/shardsplit_decoder.c

```diff
@@ -90,6 +90,45 @@ replication_origin_filter_cb(LogicalDecodingContext *ctx, RepOriginId origin_id)
 }
 
 
+/*
+ * update_replication_progress is copied from Postgres 15. We use it to send keepalive
+ * messages when we are filtering out the wal changes resulting from the initial copy.
+ * If we do not send out messages long enough, wal reciever will time out.
+ * Postgres 16 has refactored this code such that keepalive messages are sent during
+ * reordering phase which is above change_cb. So we do not need to send keepalive in
+ * change_cb.
+ */
+static void
+update_replication_progress(LogicalDecodingContext *ctx, bool skipped_xact)
+{
+#if (PG_VERSION_NUM <= PG_VERSION_15)
```

Comment:
As discussed via chat, lets ifdef this whole function and the callsite, as to make it very easy to understand in a couple of years what needs to be removed when pg15 support will drop